File: internal/cli/paths.go
package cli

import (
	"crypto/sha256"
	"fmt"
	"net/url"
	"os"
	"path/filepath"
)

// osHostname is a variable holding os.Hostname so it can be swapped in tests.
var osHostname = os.Hostname

// defaultDeviceID returns the hostname's SHA256 hash first 8 hex characters (D-033).
// The result is deterministic, anonymous, and requires zero configuration.
func defaultDeviceID() string {
	hostname, err := osHostname()
	if err != nil {
		hostname = "unknown"
	}
	h := sha256.Sum256([]byte(hostname))
	return fmt.Sprintf("%x", h[:4])
}

// ensureUserDir creates ~/.xyncra/{user_id}/{device_id}/ and returns the path.
func ensureUserDir(userID, deviceID string) (string, error) {
	home, err := os.UserHomeDir()
	if err != nil {
		return "", fmt.Errorf("ensureUserDir: %w", err)
	}
	dir := filepath.Join(home, ".xyncra", userID, deviceID)
	if err := os.MkdirAll(dir, 0700); err != nil {
		return "", fmt.Errorf("ensureUserDir: %w", err)
	}
	return dir, nil
}

// SocketPath returns the path to the Unix domain socket for the daemon.
func (c *CLIContext) SocketPath() string {
	return filepath.Join(c.UserDir, "xyncra.sock")
}

// LockPath returns the path to the lock file for the daemon.
func (c *CLIContext) LockPath() string {
	return filepath.Join(c.UserDir, "xyncra.lock")
}

// DBPathDefault returns the default path to the SQLite database.
func (c *CLIContext) DBPathDefault() string {
	return filepath.Join(c.UserDir, "xyncra.db")
}

// LogDirDefault returns the default directory for log files.
func (c *CLIContext) LogDirDefault() string {
	return filepath.Join(c.UserDir, "logs") + string(filepath.Separator)
}

// ServerURLWithUser returns the server URL with user_id and device_id query
// parameters appended. Both are required for the server to route connections
// to the correct device (D-033).
func (c *CLIContext) ServerURLWithUser() string {
	u, err := url.Parse(c.ServerURL)
	if err != nil {
		// Fallback: manual concatenation (should not happen with valid URLs).
		return c.ServerURL + "?user_id=" + url.QueryEscape(c.UserID) + "&device_id=" + url.QueryEscape(c.DeviceID)
	}
	q := u.Query()
	q.Set("user_id", c.UserID)
	q.Set("device_id", c.DeviceID)
	u.RawQuery = q.Encode()
	return u.String()
}

File: demo/web/packages/xyncra-client-cli/src/paths.ts
/**
 * Filesystem path helpers for the Xyncra CLI runtime.
 *
 * Mirrors the Go reference logic: user state lives under
 * `~/.xyncra/{userID}/{deviceID}/` and all per-runtime paths
 * (socket, lock, db, logs) are derived from that base directory.
 *
 * @module
 */

import { createHash } from 'node:crypto';
import { hostname, homedir } from 'node:os';
import { mkdirSync } from 'node:fs';
import { join } from 'node:path';

/**
 * Compute a stable device identifier from the machine hostname.
 *
 * Returns the first 8 hex characters (4 bytes) of the SHA-256 hash
 * of the hostname. Falls back to `"unknown"` when `os.hostname()` throws
 * (e.g. in some container environments).
 *
 * Matches the Go reference:
 * ```go
 * h := sha256.Sum256([]byte(hostname))
 * return fmt.Sprintf("%x", h[:4])
 * ```
 */
export function defaultDeviceID(): string {
  let host: string;
  try {
    host = hostname();
  } catch {
    host = 'unknown';
  }
  if (!host) host = 'unknown';
  const hash = createHash('sha256').update(host).digest();
  return hash.subarray(0, 4).toString('hex');
}

/**
 * Ensure the per-user/per-device directory exists and return its path.
 *
 * Creates `~/.xyncra/{userID}/{deviceID}/` with mode `0o700` if missing.
 */
export function ensureUserDir(userID: string, deviceID: string): string {
  const dir = join(homedir(), '.xyncra', userID, deviceID);
  mkdirSync(dir, { recursive: true, mode: 0o700 });
  return dir;
}

/** Path to the Unix domain socket the daemon listens on. */
export function socketPath(userDir: string): string {
  return join(userDir, 'xyncra.sock');
}

/** Path to the daemon lock file. */
export function lockPath(userDir: string): string {
  return join(userDir, 'xyncra.lock');
}

/** Default path for the local SQLite / Dexie database file. */
export function dbPathDefault(userDir: string): string {
  return join(userDir, 'xyncra.db');
}

/** Default directory for log files (trailing slash preserved). */
export function logDirDefault(userDir: string): string {
  return join(userDir, 'logs') + '/';
}

/**
 * Append `user_id` and `device_id` query parameters to a server URL.
 *
 * Uses the WHATWG URL parser so existing query strings are preserved.
 */
export function serverURLWithUser(
  serverURL: string,
  userID: string,
  deviceID: string,
): string {
  const url = new URL(serverURL);
  url.searchParams.set('user_id', userID);
  url.searchParams.set('device_id', deviceID);
  return url.toString();
}
